Distinct ref objects + destructor cause C++ codegen error (& invalid free on C backend)

### Description

I've done a rewrite of the Nim side logic of handling HDF5 identifiers in `nimhdf5` recently:

https://github.com/Vindaar/nimhdf5/pull/60

I developed the code on a devel compiler from a few weeks ago without issue. In the CI with the latest devel I suddenly got failures. Reproducing them locally was possible, but it was tricky to understand the actual issue. In my actual code sample (`tests/tattributes.nim` on HEAD~2 of the PR) using `-d:useMalloc` I could trace the problem to an invalid `free()` call from a destructor.

The C++ backend however actually produced a codegen error for the same code, which lead me to believe that indeed it is actually a compiler bug and not changed semantics of destructors (and somehow my fault).

The smallest reproducible which still gives the same C++ codegen error (but "works" fine on C backend now) is:

```nim
type
  H5IdObj = object
  H5Id = ref H5IdObj

  FileID = distinct H5Id

  H5GroupObj = object
    file_id: FileID
  H5Group = ref H5GroupObj

## This would make it work!
#proc `=destroy`*(x: FileID) = `=destroy`(cast[H5Id](x))
## If this does not exist, it also works!
proc newFileID(): FileID = FileID(H5Id())

proc `=destroy`(grp: var H5GroupObj) =
  ## Closes the group and resets all references to nil.
  if cast[pointer](grp.fileId) != nil:
    `=destroy`(grp.file_id)

let grp = H5Group()
```

Note the general setup of having a `H5Id` `ref object`, which in reality intends to wrap a HDF5 identifier. The idea is to have `distinct H5Id` types for each type of HDF5 identifiers so that we can automatically call the correct `close` procedure when the identifier goes out of scope (which is what we want in 99% of the cases). 

In this snippet it works fine if we define the `=destroy` hook for the `FileID` type (which is of course only possible since a couple of days (weeks?) now and just cast to `=destroy` of the original type. Bizarrely it also works fine if the `newFileID` proc does not exist.

### Nim Version

```
Nim Compiler Version 1.9.5 [Linux: amd64]
Compiled at 2023-06-30
Copyright (c) 2006-2023 by Andreas Rumpf

git hash: 210b10dd0d47c8c79b686a69eb8646df869cf207
active boot switches: -d:release -d:danger
```

### Current Output

```text
CC: refc_destructor_distinct.nim
/home/basti/.cache/nim/refc_destructor_distinct_d/@mrefc_destructor_distinct.nim.cpp: In function 'void eqdestroy___refc95destructor95distinct_u70(tyObject_H5GroupObj__nRZkcrXQJ43SIjRESEP8MA&)':
/home/basti/.cache/nim/refc_destructor_distinct_d/@mrefc_destructor_distinct.nim.cpp:142:60: error: cannot convert 'tyObject_H5IdObj__H6UD9cLvDyIMbqOQXxGGK1g**' to 'tyObject_H5IdObj__H6UD9cLvDyIMbqOQXxGGK1g*'
  142 |                 eqdestroy___refc95destructor95distinct_u30(&grp_p0.file_id);
      |                                                            ^~~~~~~~~~~~~~~
      |                                                            |
      |                                                            tyObject_H5IdObj__H6UD9cLvDyIMbqOQXxGGK1g**
/home/basti/.cache/nim/refc_destructor_distinct_d/@mrefc_destructor_distinct.nim.cpp:127:118: note:   initializing argument 1 of 'void eqdestroy___refc95destructor95distinct_u30(tyObject_H5IdObj__H6UD9cLvDyIMbqOQXxGGK1g*)'
  127 | N_LIB_PRIVATE N_NIMCALL(void, eqdestroy___refc95destructor95distinct_u30)(tyObject_H5IdObj__H6UD9cLvDyIMbqOQXxGGK1g* dest_p0) {
      |                                                                           ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~^~~~~~~
Error: execution of an external compiler program 'g++ -c -std=gnu++17 -funsigned-char  -w -fmax-errors=3 -fpermissive -pthread   -I/home/basti/src/nim/nim_git_repo/lib -I/home/basti/org/Misc -o /home/basti/.cache/nim/refc_destructor_distinct_d/@mrefc_destructor_distinct.nim.cpp.o /home/basti/.cache/nim/refc_destructor_distinct_d/@mrefc_destructor_distinct.nim.cpp' failed with exit code: 1
```

As mentioned above the actual code produces an invalid free call for this. I assume going by the C++ error message that the same code is generated on the C backend. But `gcc` thinks it's fine, taking the wrong pointer and thus frees the wrong memory.



### Expected Output

Should compile fine.



### Possible Solution

_No response_

### Additional Information

_No response_

--- a/compiler/semmagic.nim
+++ b/compiler/semmagic.nim
@@ -609,6 +609,14 @@ proc magicsAfterOverloadResolution(c: PContext, n: PNode,
     let op = getAttachedOp(c.graph, t, attachedDestructor)
     if op != nil:
       result[0] = newSymNode(op)
+
+      if op.typ != nil and op.typ.len == 2 and op.typ[1].kind != tyVar and
+          skipAddr(n[1]).typ.kind == tyDistinct:
+        if n[1].kind == nkSym and n[1].sym.kind == skParam and
+            n[1].typ.kind == tyVar:
+          result[1] = genDeref(n[1])
+        else:
+          result[1] = skipAddr(n[1])
   of mTrace:
     result = n
     let t = n[1].typ.skipTypes(abstractVar)

--- a/tests/destructor/tnonvardestructor.nim
+++ b/tests/destructor/tnonvardestructor.nim
@@ -193,3 +193,28 @@ block:
 
   # bug #11705
   foo()
+
+block: # bug #22197
+  type
+    H5IdObj = object
+    H5Id = ref H5IdObj
+
+    FileID = distinct H5Id
+
+    H5GroupObj = object
+      file_id: FileID
+    H5Group = ref H5GroupObj
+
+  ## This would make it work!
+  #proc `=destroy`*(x: FileID) = `=destroy`(cast[H5Id](x))
+  ## If this does not exist, it also works!
+  proc newFileID(): FileID = FileID(H5Id())
+
+  proc `=destroy`(grp: var H5GroupObj) =
+    ## Closes the group and resets all references to nil.
+    if cast[pointer](grp.fileId) != nil:
+      `=destroy`(grp.file_id)
+
+  var grp = H5Group()
+  reset(grp.file_id)
+  reset(grp)

